File: Sources/FocusCore/FocusRouter.swift
import Foundation

public struct FocusRouter: Sendable {
    public enum Decision: Equatable, Sendable {
        case allow
        case redirect(FocusEntry)
    }

    private let settings: FocusSettings
    private static let homepageHosts = Set([
        "bilibili.com",
        "www.bilibili.com",
        "m.bilibili.com",
    ])
    private static let homepagePaths = Set([
        "",
        "/",
        "/index.html",
    ])

    public init(settings: FocusSettings = .defaults) {
        self.settings = settings
    }

    public func entry(for entry: FocusEntry? = nil) -> FocusEntry {
        entry ?? settings.defaultEntry
    }

    public func entryRoute(for entry: FocusEntry? = nil) -> AppRoute {
        switch self.entry(for: entry) {
        case .dynamic, .search:
            return .dynamicFeed
        }
    }

    public func decision(for url: URL) -> Decision {
        guard settings.redirectEnabled else {
            return .allow
        }

        guard
            let host = url.host?.lowercased(),
            Self.homepageHosts.contains(host),
            Self.homepagePaths.contains(url.path)
        else {
            return .allow
        }

        return .redirect(entry())
    }

    public func redirectTarget(for url: URL) -> FocusEntry? {
        guard case let .redirect(target) = decision(for: url) else {
            return nil
        }

        return target
    }
}

File: focus-core/src/commonMain/kotlin/org/bilibilifocus/core/routing/FocusRouter.kt
package org.bilibilifocus.core.routing

import org.bilibilifocus.core.model.FocusEntry
import org.bilibilifocus.core.model.FocusSettings

class FocusRouter(private val settings: FocusSettings = FocusSettings.DEFAULTS) {

    sealed interface Decision {
        data object Allow : Decision
        data class Redirect(val entry: FocusEntry) : Decision
    }

    fun entry(entry: FocusEntry? = null): FocusEntry = entry ?: settings.defaultEntry

    fun entryRoute(entry: FocusEntry? = null): AppRoute {
        return when (this.entry(entry)) {
            FocusEntry.DYNAMIC, FocusEntry.SEARCH -> AppRoute.DynamicFeed
        }
    }

    fun decision(url: String): Decision {
        if (!settings.redirectEnabled) {
            return Decision.Allow
        }

        val parsed = BiliUrl.parse(url)
        val host = parsed.host ?: return Decision.Allow

        if (host !in homepageHosts) return Decision.Allow
        if (parsed.path !in homepagePaths) return Decision.Allow

        return Decision.Redirect(entry())
    }

    fun redirectTarget(url: String): FocusEntry? {
        return when (val decision = decision(url)) {
            is Decision.Redirect -> decision.entry
            else -> null
        }
    }

    companion object {
        private val homepageHosts = setOf("bilibili.com", "www.bilibili.com", "m.bilibili.com")
        private val homepagePaths = setOf("", "/", "/index.html")
    }
}
